### ISSUE TITLE
MapDataset.create() sets mask to False by default

### ISSUE THREAD

**Issue by AtreyeeS** [gammapy/gammapy#4207]
**Gammapy version**
`gammapy-1.0`

**Bug description**
By default, a fully empty `mask_safe` is created when making a dataset, as per
https://github.com/gammapy/gammapy/blob/647c0001a08fbda055a7aed4bf62747e08ea38e7/gammapy/datasets/map.py#L575

This is confusing, as, it leads to non-intuitive results for the background while stacking synthetic datasets, eg:

```python3
from gammapy.datasets import MapDataset
from gammapy.maps import *
import numpy as np

axis = MapAxis.from_energy_bounds("0.1 TeV", "10 TeV", nbin=3)
geom = WcsGeom.create(binsz=0.5, width=(2, 2), axes=[axis])
bkg1 = Map.from_geom(geom)
bkg1.data = np.ones(bkg1.data.shape)
cnt1 = bkg1.copy()
dataset1 = MapDataset(counts=cnt1, background=bkg1, name="dataset-1")

stacked = MapDataset.from_geoms(**dataset1.geoms)
for i in range(3):
    stacked.stack(dataset1)
    print(stacked.counts.data.sum(), stacked.background.data.sum())
```

Output being:

```
48.0 48.0
96.0 48.0
144.0 48.0
```

Since `MapDatasetMaker.run()` creates a `True` `mask_safe` by default, there should be no erroneous results for users going through the standard data reduction loop. Still, I think it would be good to be consistent and set the mask to True in all cases by default.

**Comment by adonath** [comment]
Thanks @AtreyeeS! 

> Since MapDatasetMaker.run() creates a True mask_safe by default, there should be no erroneous results for users going through the standard data reduction loop. Still, I think it would be good to be consistent and set the mask to True in all cases by default.

This is not a bug, but the convention we introduced for `MapDataset.create()` and `.from_geoms()`. The `.mask_safe` will be set to all false by default. This is needed because while stacking, masks are combined using a logical or. If we started with a mask set all true, stacking could never modify the mask: it would just remain all true. Which is even worse default behavior I think.


However looking at `MapDataset.stack()` in noticed indeed one inconsistency. For some reason the background stacking applies the `self.mask_safe` while all the others quantities do not. So I think the actual bug here is the multiplication with `self.mask_safe` here: https://github.com/gammapy/gammapy/blob/main/gammapy/datasets/map.py#L786.

**Comment by adonath** [comment]
The modified example following the convention, works as expected:

```
from gammapy.datasets import MapDataset
from gammapy.maps import *
import numpy as np

axis = MapAxis.from_energy_bounds("0.1 TeV", "10 TeV", nbin=3)
geom = WcsGeom.create(binsz=0.5, width=(2, 2), axes=[axis])
bkg1 = Map.from_geom(geom)
bkg1.data = np.ones(bkg1.data.shape)
cnt1 = bkg1.copy()

mask_safe = Map.from_geom(geom, data=True)
dataset1 = MapDataset(
    counts=cnt1,
    background=bkg1,
    mask_safe=mask_safe,
    name="dataset-1"
)

stacked = MapDataset.from_geoms(**dataset1.geoms)
for i in range(3):
    stacked.stack(dataset1)
    print(stacked.counts.data.sum(), stacked.background.data.sum())
```
Prints:

```
48.0 48.0
96.0 96.0
144.0 144.0
```

**Comment by AtreyeeS** [comment]
Thanks for the clarification @adonath 

> This is not a bug, but the convention we introduced for MapDataset.create() and .from_geoms(). The .mask_safe will be set to all false by default. This is needed because while stacking, masks are combined using a logical or.

Indeed I remember now! This was discussed in a dev call in details

> The modified example following the convention, works as expected:

Of course. This is not a blocker by any means, but can be a source of confusion. We do not specify anywhere that it is necessary to pass a `mask_safe` while creating a dataset from some previously computed maps.
In the absence of `mask_safe` on some datasets, the result of the stacking depends upon the order in which the datasets are passed, which is weird because stacking should be commutative.
Its a case of silently giving wrong results, that can be rather annoying to understand later...

> So I think the actual bug here is the multiplication with self.mask_safe

Agreed! Off hand, I do not see a specific reason why the background stacking should need to consider the safe mask.
Probably it is good to remove this then?

**Comment by registerrier** [comment]
> > So I think the actual bug here is the multiplication with self.mask_safe
> 
> Agreed! Off hand, I do not see a specific reason why the background stacking should need to consider the safe mask. Probably it is good to remove this then?

The masking is required to avoid including counts excluded from the other dataset. It has to be applied equivalently for background npred. 
Now the multiplication by `self.mask_safe` is done only for the background, which is surprising. Only the valid counts and background should be stacked.

**Comment by adonath** [comment]
@registerrier At some point the implementation of `.stack()` was simplified to only take into account the mask of the `other`, because in 90% of the cases one stack into an empty dataset anyway. In case the dataset is not empty one can rely on the pattern:

```python3
stacked = stacked.to_masked()
```

That's why I was surprised to see that the `self.mask_safe` was still applied. However the recommended patterns are not affected by this...so if at all it is only a very minor bug.

**Comment by registerrier** [comment]
> @registerrier At some point the implementation of `.stack()` was simplified to only take into account the mask of the `other`, because in 90% of the cases one stack into an empty dataset anyway. In case the dataset is not empty one can rely on the pattern:
> 
> ```python
> stacked = stacked.to_masked()
> ```
But is it equivalent? Here, the stacked mask (which is larger than the initial one) is applied to stacked events, but the latter contain all events of the initial dataset.
It should not be the same as stacking only events in the respective masks. 

For the typical use case where we iteratively stack onto the target dataset, this works perfectly fine, indeed. 
But I suppose this causes the non-commutativity of `stack` that @AtreyeeS is referring  to.

**Comment by adonath** [comment]
@registerrier Just clarify, the masking should happened before stacking:

Stack into empty dataset:
```python3

stacked = MapDataset.from_geoms(**datasets[0].geoms)

for dataset in datasets:
    stacked.stack(dataset)

```

Stack into non-empty dataset:

```python3

stacked = datasets[0]

stacked = stacked.to_masked()

for dataset in datasets[1:]:
    stacked.stack(dataset)

```
Leaving aside cutout handling this should give equivalent results...

**Comment by AtreyeeS** [comment]
This issue was reported to me by a user who was stacking some Fermi datasets, and ended up with exceptionally high significances after stacking. He was following our Fermi-LAT notebook to create the dataset, where, we indeed do not mention anything about masks. The simplest solution is to add in the tutorial, but I wanted to bring this up for discussion because this may have effects in other science cases...

**Comment by AtreyeeS** [comment]
Maybe `MapDataset.mask_safe` can be automatically set as `True` on init if `mask_safe` is `None` ? `MapDataset.from_geoms` specifically creates an empty mask, so there would be no changes to the current implementation.

**Comment by adonath** [comment]
> Maybe MapDataset.mask_safe can be automatically set as True on init if mask_safe is None ? MapDataset.from_geoms specifically creates an empty mask, so there would be no changes to the current implementation.

`mask_safe = None` is already handled like this, it behaves as if the mask was all true.

**Comment by registerrier** [comment]
This is now corrected in #4275 . Closing.
